securesystemslib/storage.py

```diff
@@ -0,0 +1,219 @@
+
+import abc
+import errno
+import logging
+import os
+import shutil
+
+import six
+
+import securesystemslib.exceptions
+
+logger = logging.getLogger(__name__)
+
+if six.PY2:
+  FileNotFoundError = OSError # pragma: no cover
+
+
+
+class StorageBackendInterface():
+  """
+  <Purpose>
+  Defines an interface for abstract storage operations which can be implemented
+  for a variety of storage solutions, such as remote and local filesystems.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+
+  @abc.abstractmethod
+  def get(self, filepath):
+    """
+    <Purpose>
+      Retrieve a file from the storage backend.
+
+    <Arguments>
+      filepath:
+        The full path of the file to be retrieved.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file does not exist or is
+      no accessible.
+
+    <Returns>
+      A file-like object of the file at 'filepath'.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def put(self, fileobj, filepath):
+    """
+    <Purpose>
+      Store a file-like object in the storage backend.
+
+    <Arguments>
+      fileobj:
+        The file-like object to be stored.
+
+      filepath:
+        The full path to the location where 'fileobj' will be stored.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file can not be stored.
+
+    <Returns>
+      None
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def getsize(self, filepath):
+    """
+    <Purpose>
+      Retrieve the size, in bytes, of the file at 'filepath'.
+
+    <Arguments>
+      filepath:
+        The full path to the file.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file does not exist or is
+      not accessible.
+
+    <Returns>
+      The size in bytes of the file at 'filepath'.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def create_folder(self, filepath):
+    """
+    <Purpose>
+      Create a folder at filepath and ensure all intermediate components of the
+      path exist.
+
+    <Arguments>
+      filepath:
+        The full path of the folder to be created.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the folder can not be
+      created.
+
+    <Returns>
+      None
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def list_folder(self, filepath):
+    """
+    <Purpose>
+      List the contents of the folder at 'filepath'.
+
+    <Arguments>
+      filepath:
+        The full path of the folder to be listed.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file does not exist or is
+      not accessible.
+
+    <Returns>
+      A list containing the names of the files in the folder. May be an empty
+      list.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+
+
+
+class FilesystemBackend(StorageBackendInterface):
```

Comment:
Should this be FileSystemBackend?

Reply:
I've seen file system and filesystem used, I don't have a strong preference. I'd be happy to change it.

Reply:
I have a _slight_ preference for `FilesystemBackend` based on "filesystem" being a full noun, but this backend doesn't affect me in either case 😉